### ISSUE TITLE
rpcclient: Add estimatesmartfee support

### ISSUE THREAD

**Issue by gracenoah** [btcsuite/btcd#1146]
The JSON-RPC method `estimatesmartfee` was added to bitcoind on version 0.12 (https://github.com/bitcoin/bitcoin/pull/6134) but there is currently no way of reaching it through btcd.

**Comment by hiromaily** [comment]
+1

**Comment by hiromaily** [comment]
`EstimateFee()` returns error.
```
-32: estimatefee is deprecated and will be fully removed in v0.17. To use estimatefee in v0.16, restart bitcoind with -deprecatedrpc=estimatefee.
Projects should transition to using estimatesmartfee before upgrading to v0.17
```
So this should be fixed.

**Comment by johandroz** [comment]
Got the same issue. You can use a RawRequest to get the `estimatesmartfee`.
I did something like:
```go

// RPCClient is the interface of the RPC client.
type RPCClient interface {
	RawRequest(method string, params []json.RawMessage) (json.RawMessage, error)
}

// EstimateMode encodes the possible estimation modes of estimatesmartfee
type EstimateMode string

const (
	// UnsetEstimeMode identifies the UNSET estimation strategy used by estimatesmartfee
	UnsetEstimeMode EstimateMode = "UNSET"
	// ConservativeEstimeMode identifies the CONSERVATIVE estimation strategy used by estimatesmartfee
	ConservativeEstimeMode EstimateMode = "CONSERVATIVE"
	// EconomicalEstimeMode identifies the ECONOMICAL estimation strategy used by estimatesmartfee
	EconomicalEstimeMode EstimateMode = "ECONOMICAL"
)

// EstimateSmartFeeReply is the result of the EstimateSmartFee call.
type EstimateSmartFeeReply struct {
	FeeRate float64 `json:"feerate"`
	Blocks  int64   `json:"blocks"`
}

// EstimateSmartFee estimates the approximate fee per kilobyte needed for a transaction to begin
// confirmation within 'numBlocks' blocks if possible and return the number of blocks
// for which the estimate is valid.
// The fee estimate ("ECONOMICAL", "CONSERVATIVE" or "" (default to CONSERVATIVE)) mode defines whether
// to return a more conservative estimate which also satisfies a longer history. A conservative estimate
// potentially returns a higher feerate and is more likely to be sufficient for the desired target, but
// is not as responsive to short term drops in the prevailing fee market.
func EstimateSmartFee(client RPCClient, numBlocks int64, mode EstimateMode) (*EstimateSmartFeeReply, error) {
	var params []json.RawMessage
	{
		numBlocksJSON, err := json.Marshal(numBlocks)
		if err != nil {
			return nil, errors.Wrap(err, "could not marshal numBlocks")
		}
		modeJSON, err := json.Marshal(mode)
		if err != nil {
			return nil, errors.Wrap(err, "could not marshal mode")
		}
		params = []json.RawMessage{numBlocksJSON, modeJSON}
	}

	var data json.RawMessage
	{
		var err error
		data, err = client.RawRequest("estimatesmartfee", params)
		if err != nil {
			return nil, errors.Wrap(err, "could not estimate fee")
		}
	}

	var reply = struct {
		FeeRate float64  `json:"feerate"`
		Blocks  int64    `json:"blocks"`
		Errors  []string `json:"errors,omitempty"`
	}{}
	var err = json.Unmarshal(data, &reply)
	if err != nil {
		return nil, errors.Wrap(err, "could not unmarshal estimate smart fee result")
	}

	if len(reply.Errors) != 0 {
		return nil, errors.New(strings.Join(reply.Errors, ","))
	}

	return &EstimateSmartFeeReply{
		Blocks:  reply.Blocks,
		FeeRate: reply.FeeRate,
	}, nil
}
```

**Comment by gracenoah** [comment]
I recommend just using https://github.com/stevenroose/go-bitcoin-core-rpc instead.

**Comment by onyb** [comment]
Fixed in 8b1be46.

**Comment by jcvernaleo** [comment]
Closing as complete.
